fdb-record-layer-geophile/src/main/java/com/apple/foundationdb/record/geophile/SpatialIndexValidator.java

```diff
@@ -0,0 +1,46 @@
+/*
+ * SpatialIndexValidator.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2019 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.geophile;
+
+import com.apple.foundationdb.annotation.API;
+import com.apple.foundationdb.record.metadata.Index;
+import com.apple.foundationdb.record.metadata.IndexValidator;
+import com.apple.foundationdb.record.metadata.MetaDataValidator;
+
+import javax.annotation.Nonnull;
+
+/**
+ * Validate that the key expression in a spatial index is valid.
+ */
+@API(API.Status.EXPERIMENTAL)
+class SpatialIndexValidator extends IndexValidator {
+    public SpatialIndexValidator(Index index) {
+        super(index);
+    }
+
+    @Override
+    public void validate(@Nonnull MetaDataValidator metaDataValidator) {
+        super.validate(metaDataValidator);
+        validateNotGrouping();
+        validateNotVersion();
+        // TODO: Validate something like that bottom-right child is SpatialFunctionKeyExpression.
```

Comment:
Since the index maintainer now needs to get the spatial expression to get the space, I made that the validator. The extraction code isn't very general; it feels like something in the core ought to be able to do this.

Reply:
Okay, yeah, that seems like the right place to put the validator, with a comment left on the validator itself. Agreed that this seems like something we care about more generally, so the key expression manipulation tools should be able to handle it with fewer seams. Hmm.